### PR TITLE
[fix] OSS dict: simplified logic, bugfix for interleaved param groups 

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by blefaudeux** [comment]
Pytorch1.5 needs fixing, as always..

File: fairscale/optim/oss.py [hunk 1]
```diff
@@ -160,15 +159,6 @@ def local_params(self) -> List[torch.Tensor]:
         # Make sure that the iterator is not consumed, only expose a copy
         return self._local_params
 
-    @property
-    def index_to_param(self) -> Dict[int, torch.Tensor]:
```
Comment (blefaudeux) [inline fairscale/optim/oss.py:164]: this was an optimisation for very big models, which I think was a mistake: this increased the code footprint, and really, really big models will not be trained with OSS anyway. On top of that, this is not needed with this PR except for torch 1.5, which will be deprecated soonish

File: tests/optim/test_oss.py [hunk 2]
```diff
@@ -811,9 +817,11 @@ def check_optimizer_equivalence(optimizer: Type[torch.optim.Optimizer], change_t
 
         # Define a model to be trained by OSS
         oss_module = torch.nn.Sequential(trunk, head)
+
+        # Make sure that the param groups are interleaved, to catch an ordering bug in the state dict
         oss_trainable_params = [
-            {"params": trunk.parameters(), "lr": 1e-5},
```
Comment (blefaudeux) [inline tests/optim/test_oss.py:815]: this unit test let a "subtle" bug pass: it was testing param groups which are sequential with respect to the model parameters, turns out that there was a bug if this was not the case.

File: fairscale/optim/oss.py [hunk 3]
```diff
@@ -389,17 +379,34 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
                 from a call to :meth:`state_dict`
         """
 
-        # NOTE: PyTorch 1.5 does not index linearly but with the id(params) at saving time
-        # we work around that here by using the fact that the params are ordered as in the param_groups
-        pytorch15_index_redirect = {k: i for i, k in enumerate(state_dict["state"].keys())}
```
Comment (blefaudeux) [inline fairscale/optim/oss.py:394]: this was the arguably a little subtle bug: this implied that the loaded state dict and the param groups were sorted in the same way, which is not always the case actually (happened to be the case for the unit test in place..)

File: fairscale/optim/oss.py [hunk 4]
```diff
@@ -389,17 +379,34 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
...
+        # Update the state, trusting the ordering in param_groups
+        # Apart from the removal of states not owned by this rank, the pytorch logic is kept
+        # (See torch.optim.optimizer)
+        id_map = {
```
Comment (blefaudeux) [inline fairscale/optim/oss.py:385]: this is exactly the same logic as pytorch (as commented), which I think is a safer and sane-er (if this word exist ?) way to go

File: fairscale/optim/oss.py [hunk 5]
```diff
@@ -389,17 +379,34 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
...
+            old_id: p
+            for old_id, p in zip(
+                chain.from_iterable((g["params"] for g in state_dict["param_groups"])),
+                chain.from_iterable((g["params"] for g in self.param_groups)),
+            )
+        }
+
+        # FIXME: pytorch1.5 compatibility, to be removed when 1.5 support ends
+        _param_list = list(chain.from_iterable((g["params"] for g in self.param_groups)))
```
Comment (blefaudeux) [inline fairscale/optim/oss.py:394]: the assumption (same as pytorch) is not that param_groups are ordered in the same way in between checkpoint and current state. Previous assumption was that checkpointed *state* and current param_groups are ordered in the same way, which turned out to be false in some cases. cc @vedanuj

**Comment by blefaudeux** [comment]
ping reviewers

**Review by min-xu-ai — APPROVED** [review]
lgtm

**Pull request merged**